src/coreComponents/physicsSolvers/wavePropagation/AcousticWaveEquationSEM.cpp

```diff
@@ -0,0 +1,821 @@
+  {
+    elemRegion.forElementSubRegionsIndex< CellElementSubRegion >( [&]( localIndex const,
+                                                                       CellElementSubRegion & elementSubRegion )
+    {
+      arrayView2d< localIndex const, cells::NODE_MAP_USD > const & elemsToNodes = elementSubRegion.nodeList();
+
+      finiteElement::FiniteElementBase const &
+      fe = elementSubRegion.getReference< finiteElement::FiniteElementBase >( getDiscretizationName() );
+      finiteElement::dispatch3D( fe,
+                                 [&]
+                                   ( auto const finiteElement )
+      {
+        using FE_TYPE = TYPEOFREF( finiteElement );
+
+        constexpr localIndex numNodesPerElem = FE_TYPE::numNodes;
+        constexpr localIndex numQuadraturePointsPerElem = FE_TYPE::numQuadraturePoints;
+
+        real64 N[numNodesPerElem];
+        real64 gradN[ numNodesPerElem ][ 3 ];
+
+        /* Unit test
+           // For unit test, intialize p_n with x-coordinate of mesh vertices {(x,y,z), x,y,z \in R}
+           for( localIndex a=0; a<nodes.size(); ++a )
+           {
+            p_n[a] =  X(a,0); //1.0; // X(a,0);
+           }
+         */
+
+        for( localIndex k=0; k < elemsToNodes.size( 0 ); ++k )
+        {
+          real64 xLocal[numNodesPerElem][3];
+
+          for( localIndex a=0; a< numNodesPerElem; ++a )
+          {
+            for( localIndex i=0; i<3; ++i )
+            {
+              xLocal[a][i] = X( elemsToNodes( k, a ), i );
+            }
+          }
+
+          for( localIndex q=0; q<numQuadraturePointsPerElem; ++q )
+          {
+            FE_TYPE::calcN( q, N );
+
+            real64 const detJ = finiteElement.template getGradN< FE_TYPE >( k, q, xLocal, gradN );
+
+            for( localIndex i=0; i<numNodesPerElem; ++i )
+            {
+              for( localIndex j=0; j<numNodesPerElem; ++j )
+              {
+                real64 Rh_ij = 0.0;
+                for( localIndex a=0; a < 3; ++a )
+                {
+                  Rh_ij +=  detJ * gradN[i][a]*gradN[j][a];
+                }
+
+                stiffnessVector[elemsToNodes[k][i]] += Rh_ij*p_n[elemsToNodes[k][j]];
+              }
+
+            }
+
+          }
+
+        }
+        /* Unit test
+           // compute <\nabla p, \nabla p> = <p, Rh p>
+           real64 prodScalar = 0.0;
+           for( localIndex a=0; a<nodes.size(); ++a )
+           {
+            prodScalar +=stiffnessVector[a]*p_n[a];
+           }
+           std::cout << " Scalar product p_n and stiffnessVector = " << prodScalar << std::endl;
+           GEOSX_ERROR_IF( true, " Stop test Stiffness matrix Ok" );
+         */
+      } );
+    } );
+  } );
+
+  addSourceToRightHandSide( time_n, rhs );
+
+  /// Calculate your time integrators
+  real64 dt2 = dt*dt;
+  for( localIndex a=0; a<nodeManager.size(); ++a )
+  {
+    if( freeSurfaceNodeIndicator[a]!=1 )
```

Comment:
Ok, I see where the `freeSurfaceNodeIndicator` is used now. In theory, skipping these calculations may not be mandatory, since you apply the free surface BC after. I would keep this check here, in the next PR(s) we can make some tests and make sure that this check is useful

Reply:
Right now, I call the function applyFreeSurfaceBC once to set the arrays freeSurfaceNodeIndicator and freeSurfaceFaceIndicator before computing the damping matrix. The only way to keep the pressure value at zero in this version is to do this check so I don't need to call applyFreeSurfaceBC at each time step.